x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/bulk/TransportBulkShardOperationsAction.java

```diff
@@ -131,4 +131,23 @@ protected BulkShardOperationsResponse newResponseInstance() {
         return new BulkShardOperationsResponse();
     }
 
+    /**
+     * Custom write result to include global checkpoint after ops have been replicated.
+     */
+    static class CcrWritePrimaryResult extends WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> {
+
+        CcrWritePrimaryResult(BulkShardOperationsRequest request, Translog.Location location, IndexShard primary, Logger logger) {
+            super(request, new BulkShardOperationsResponse(), location, null, primary, logger);
+        }
+
+        @Override
+        protected void respondIfPossible(Exception ex) {
+            assert Thread.holdsLock(this);
+            // maybe invoked multiple times, but that is ok as global checkpoint does not go backwards
```

Comment:
I wonder if we should override the respond method - this one is called once while the replication operation is completed (but post writes may still be inflight). Seems like cleaner logic to follow and we can write a comment there as the why we do it (get a fresh global checkpoint once the current batch has been fully replicated)